vllm/model_executor/layers/sampler.py

```diff
@@ -139,6 +142,35 @@ def _get_bin_counts_and_mask(
     return bin_counts, mask
 
 
+def _apply_min_tokens_penalty(
```

Comment:
This looks great @tjohnson31415. But I think technically the token_ids_to_penalize should be determined per seq_group (i.e. also within the loop) since they may be different per seq group. The indexing gets a bit tricker but I think it might be possible with `scatter_` with `src=-torch.inf`. Or else could group the sequences that share the same list of tokens to pernalize.

Reply:
Heh, yup. Thanks for pointing that out. I still need to write some tests for this 😅. I pushed a fix to build a list of coordinates to penalize within the loop so the stop ids are per seq_group.

I was trying to use `scatter` initially, but couldn't figure out how to get it to work. In particular, the scatter uses a rectangular tensor and doesn't seem to have a way to "skip" rows where we don't want to scatter into. So I think a gather-modify-scatter (where we gather across all sequences and stop token ids) would work, but we'd still need to index into the gather'd tensor to set the `-inf` values.